### PR TITLE
doc: Understanding RisingWave Macros

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As per title.

Include 2 examples: `commit_meta!` and `#[aggregate]`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/README.md [hunk 1]
```diff
@@ -17,3 +17,57 @@ Almost all components of RisingWave are developed in rust, and they are split to
...
+
+To understand it, first I dump the expanded code to a file:
+
+```bash
+cargo expand -p risingwave_expr > expr.rs
+```
+
+Then we identify the `#[aggregate]` macro call to examine. In this case it's `string_agg`.
+
+First, we know that Rust merges the individual modules into one big file, namespaced by the module name.
```
Comment (wangrunji0408) [inline src/README.md:64]: I guess the merge is done by `cargo expand`?
```suggestion
First, we know that `cargo expand` merges the individual modules into one big file, namespaced by the module name.
```

File: src/README.md [hunk 2]
```diff
@@ -17,3 +17,57 @@ Almost all components of RisingWave are developed in rust, and they are split to
...
+
+In this case `string_agg` is its own module, so we can search for `mod string_agg` in the `expr.rs` we just generated.
+
+We can see that the expanded `string_agg` now also contains `extern fn string_agg_varchar_varchar_varchar`.
+
+Reading the code we understand now that it does the following:
+1. Register it to a global registry for aggregates.
+2. "it" here refers to an `AggFuncSig`, whose definition we can checkout.
+3. Of note is that it will generate an anonymous struct which implements the `AggFunc` trait.
```
Comment (wangrunji0408) [inline src/README.md:73]:
```suggestion
3. Of note is that it will generate an anonymous struct which implements the `AggregateFunction` trait.
```

**Review by wangrunji0408 — APPROVED** [review]
LGTM! Thanks for this work to improve developer experience!

File: src/README.md [hunk 3]
```diff
@@ -17,3 +17,57 @@ Almost all components of RisingWave are developed in rust, and they are split to
 [state-store]: https://github.com/risingwavelabs/risingwave/blob/main/docs/state-store-overview.md
 [meta-service]: https://github.com/risingwavelabs/risingwave/blob/main/docs/meta-service.md
 [risedev]: https://github.com/risingwavelabs/risingwave/tree/main/src/risedevtool
+
+## Understanding RisingWave Macros
+
+RisingWave uses a lot of macros to simplify development.
+You may choose the read to macro definition directly,
+but something complementary / easier is to use `cargo expand`
```
Comment (xxchan) [inline src/README.md:25]: even more handy is https://rust-analyzer.github.io/manual.html#expand-macro-recursively

image

Comment (kwannoel) [reply src/README.md:25]: Thanks!

**Pull request merged**